sql-parser/src/main/java/io/crate/sql/tree/ArraySubQueryExpression.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.sql.tree;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public class ArraySubQueryExpression extends Expression {
+
+    @Nullable
+    private final Expression subqueryExpression;
+
+    public ArraySubQueryExpression(@Nullable Expression subqueryExpression) {
```

Comment:
I think it should be possible to remove the `Nullable` annotation?

Reply:
Would it make sense to be more restrictive here and only allow `SubqueryExpression` instances? Given that the name of this class `ArraySubQueryExpression` implies that only subqueries are allowed.

Reply:
good point!

Reply:
Thing is that if I use the specific type in the constructor, then I need to cast the visit return type to that specific type in the ASTBuilder: 
```
    @Override
    public Node visitArraySubquery(SqlBaseParser.ArraySubqueryContext ctx) {
        return new ArraySubQueryExpression((SubqueryExpression) visit(ctx.subqueryExpression()));
    }
```
where in all other case the more general class Expression is being used.

What do you think is better? I would go for the generic type Expression

Reply:
Although consistency is nice, I think it's more important to catch errors early and make it impossible to construct invalid objects. .E.g. creating a `ArraySubQueryExpression` with a literal inside doesn't make any sense. And it would move the validation/error handling to all consumers of the `ArraySubQueryExpression`

So I'd be in favor of casting to `SubqueryExpression`